sql/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteThreadRunner.scala

```
@@ -63,6 +63,16 @@ private[connect] class ExecuteThreadRunner(executeHolder: ExecuteHolder) extends
     }
   }
 
+  /**
+   * Checks if the execution is completed.
+   *
+   * @return
+   *   true if the execution is completed.
+   */
+  private[connect] def isCompleted(): Boolean = {
+    state.getAcquire() == ThreadState.completed
+  }
```

This still depends on our code reaching and executing the state transition into completed... How does it help for "ExecuteThreadRunner fails to record the completion/error to the observer (e.g., due to OOM)" where I understand the thread would just die / not reach the transition?
Edit: I think I got it, it's about errors on the RPC sender i.e. handler side, not execution thread side.

---

Yes, the sender detects the state of the execute thread runner and acts accordingly.

---

What if the `ExecuteThreadRunner` just dies without executing the transition to ThreadState.completed?

---

The order of read operations is 1. executionCompleted (load-acquire) -> 2. { sizeLimitReached || deadlineReached }. 

After 1., there are two possible states.
1. Either completed or an error is recorded in the observer: gotResponse must be "true" so the entire block won't be executed. In other words, if something's been consumed, the new code block in this change won't be executed.
2. None recorded: since !sizeLimitReached && !gotResponse, there is no way further data will be produced by any executors for the query, therefore raising an error will be the correct action.

But, thinking about it a little bit further, there can be an unfortunate occasion where deadlineReached becomes true just before entering the while loop (line 237), e.g., when the OS scheduler kicks in.
-> Hm...

---

If ExecuteThreadRunner fails to start, an exception will be raised from SparkConnectExecutePlanHandler.

(disclaimer: I'm not an JVM expert) Can there be any situation where a thread is silently gone?

---

hm, I guess it should always result in an exception coming into the catch all `finally` that we have, unless the whole Java process is just force killed by the OS, in which case we don't have to care about other threads either....